src/highlight/highlightUtil.ts

```diff
@@ -58,3 +58,56 @@ export const centerHighlight = (shapes: Shape[]): Position => {
 export function isHighlight(annotation: Annotation): annotation is AnnotationHighlight {
     return annotation?.target?.type === Type.highlight;
 }
+
+export const dedupeRects = (rects: Shape[], threshold = 0.1): Shape[] => {
+    const dedupedRects: Shape[] = [];
+
+    rects.forEach(curr => {
+        const prev = dedupedRects.pop();
+        // empty list, push current
+        if (!prev) {
+            dedupedRects.push(curr);
+            // The return in forEach callback is acting like a continue in for loop
+            return;
+        }
+
+        const heightThreshold = Math.max(Math.round(prev.height * threshold), Math.round(curr.height * threshold), 1);
+
+        if (prev.x === curr.x && prev.width === curr.width && Math.abs(prev.y - curr.y) <= heightThreshold) {
```

Comment:
Should we worry about the `x` & `width` varying by some threshold as well? How does this perform across browsers?

Reply:
Firefox has no dups, amazing! Safari's dups' widths are a little different. I'll add threshold to x and remove width in the clause but use width together with height to calculate area in the if body.